### PR TITLE
Handling no-return types

### PR DESCRIPTION
**Pull request by madmann91**
I am not sure about how to handle the no-return type. This fix makes 23 test cases pass. Please correct any mistake or stylistic issue if you see one.

### CODE REVIEW

File: src/impala/sema/infersema.cpp [hunk 1]
```diff
@@ -135,6 +135,9 @@ class InferSema : public TypeTable {
             return t->rebuild(nargs);
         }
 
+        if (u->isa<NoRetType>()) return unify(representative(t), representative(u))->type;
+        if (t->isa<NoRetType>()) return unify(representative(u), representative(t))->type;
+
```
Comment (leissa) [inline src/impala/sema/infersema.cpp]: I don't think this is a good idea:
Say, ```t``` is ```noret ``` and ```u``` is ```int```. Then, this code will unify ```t``` and ```u``` - but I don't think those type are supposed to be the same thing.

**Comment by madmann91** [comment]
But then what do we do with this code:

```
fn f() -> int {
    if (true) {
        return(4)
    } else {
        5
    }
}
```

This is taken from the tests (sema/positive/if03.impala).

**Comment by madmann91** [comment]
My impression was that, because of this, whenever something is `noret`, then unification returns `noret`.

**Comment by leissa** [comment]
The problem is the following: If you unfiy ```float``` with ```noret``` and also ```int``` with ```noret``` both ```float``` and ```int``` are considered the same type - which is of course broken. On the other hand it's still unclear how to resolve this issue in a clean way. I'll talk with Klaas about this issue. Hopefully, we can come up with a solution by the end of the week.

**Comment by madmann91** [comment]
Can't we make `noret` polymorphic? That would mean unifying `noret` and `int` would only make `int` the same type as a particular `noret`. The problem in the current implementation is that `noret` is treated like a type, although it behaves more like a type variable.

**Comment by leissa** [comment]
That's actually a nice idea. We'll discuss that.

**Comment by leissa** [comment]
This should do the trick - we don't unify at all when we encounter ```noret```. We simply return the other type.

**Pull request merged**